src/bgw/job_stat_history.c

```diff
@@ -0,0 +1,168 @@
+		.table = catalog_get_table_id(catalog, BGW_JOB_STAT_HISTORY),
+		.index = catalog_get_index(catalog, BGW_JOB_STAT_HISTORY, indexid),
+		.nkeys = nkeys,
+		.scankey = scankey,
+		.flags = SCANNER_F_KEEPLOCK,
+		.tuple_found = tuple_found,
+		.filter = tuple_filter,
+		.data = data,
+		.lockmode = lockmode,
+		.scandirection = ForwardScanDirection,
+	};
+
+	return ts_scanner_scan_one(&scanctx, false, "bgw job stat");
+}
+
+static inline bool
+bgw_job_stat_history_scan_id(int64 bgw_job_history_id, tuple_found_func tuple_found,
+							 tuple_filter_func tuple_filter, void *data, LOCKMODE lockmode)
+{
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_bgw_job_stat_history_pkey_idx_id,
+				BTEqualStrategyNumber,
+				F_INT8EQ,
+				Int64GetDatum(bgw_job_history_id));
+
+	return bgw_job_stat_history_scan_one(BGW_JOB_STAT_HISTORY_PKEY_IDX,
+										 scankey,
+										 1,
+										 tuple_found,
+										 tuple_filter,
+										 data,
+										 lockmode);
+}
+
+static ScanTupleResult
+bgw_job_stat_history_tuple_mark_end(TupleInfo *ti, void *const data)
+{
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	BgwJobStatHistoryContext *context = (BgwJobStatHistoryContext *) data;
+
+	Datum values[Natts_bgw_job_stat_history] = { 0 };
+	bool nulls[Natts_bgw_job_stat_history] = { 0 };
+	bool doReplace[Natts_bgw_job_stat_history] = { 0 };
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_pid)] = Int32GetDatum(MyProcPid);
+	doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_pid)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_execution_finish)] =
+		TimestampTzGetDatum(ts_timer_get_current_timestamp());
+	doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_execution_finish)] = true;
+
+	values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_succeeded)] =
+		BoolGetDatum((context->result == JOB_SUCCESS));
+	doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_succeeded)] = true;
+
+	if (context->edata != NULL)
+	{
+		values[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_error_data)] =
+			JsonbPGetDatum(context->edata);
+		doReplace[AttrNumberGetAttrOffset(Anum_bgw_job_stat_history_error_data)] = true;
+	}
+
+	HeapTuple new_tuple =
+		heap_modify_tuple(tuple, ts_scanner_get_tupledesc(ti), values, nulls, doReplace);
+
+	ts_catalog_update(ti->scanrel, new_tuple);
+
+	heap_freetuple(new_tuple);
+
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+void
+ts_bgw_job_stat_history_mark_end(BgwJob *job, JobResult result, Jsonb *edata)
+{
+	/* in case of the GUC is false and a failure occurs then we need to force mark the start of the
+	 * job execution */
+	if (!ts_guc_enable_job_execution_logging && result != JOB_SUCCESS)
+	{
+		ts_bgw_job_stat_history_mark_start(job);
+
+		/* the following mark end should see the inserted history job */
+		CommandCounterIncrement();
+	}
+
+	BgwJobStatHistoryContext context = {
```

Comment:
Couldn't we still run into the problem of having an invalid job ID at this point (i.e., `ts_guc_enable_job_execution_logging` is modified between job start and end)?

Reply:
Fixed!